src/Microsoft.Maintainability.Analyzers/Core/DoNotIgnoreMethodResults.cs

```diff
@@ -1,68 +1,146 @@
+                    {
+                        case OperationKind.ObjectCreationExpression:
+                            var ctor = ((IObjectCreationExpression)expression).Constructor;
+                            if (ctor != null)
+                            {
+                                messageFormat = MicrosoftMaintainabilityAnalyzersResources.DoNotIgnoreMethodResultsMessageObjectCreation;
+                                targetMethodName = ctor.ContainingType.Name;
+                            }
+                            break;
+
+                        case OperationKind.InvocationExpression:
+                            var invocationExpression = ((IInvocationExpression)expression);
+                            var targetMethod = invocationExpression.TargetMethod;
+                            if (targetMethod == null)
+                            {
+                                break;
+                            }
+
+                            if (IsStringCreatingMethod(targetMethod))
+                            {
+                                messageFormat = MicrosoftMaintainabilityAnalyzersResources.DoNotIgnoreMethodResultsMessageStringCreation;
+                            }
+                            else if (IsTryParseMethod(targetMethod))
+                            {
+                                messageFormat = MicrosoftMaintainabilityAnalyzersResources.DoNotIgnoreMethodResultsMessageTryParse;
+                            }
+                            else if (IsHResultOrErrorCodeReturningMethod(targetMethod))
+                            {
+                                messageFormat = MicrosoftMaintainabilityAnalyzersResources.DoNotIgnoreMethodResultsMessageHResultOrErrorCode;
+                            }
+
+                            targetMethodName = targetMethod.Name;
+                            break;
+                    }
+
+                    if (messageFormat != null)
+                    {
+                        var message = string.Format(messageFormat, method.Name, targetMethodName);
+                        var diagnostic = Diagnostic.Create(Rule, expression.Syntax.GetLocation(), message);
+                        opContext.ReportDiagnostic(diagnostic);
+                    }
+                }, OperationKind.ExpressionStatement);
+            });
+        }
+
+        private static bool IsStringCreatingMethod(IMethodSymbol method)
+        {
+            return method.ContainingType.SpecialType == SpecialType.System_String &&
+                s_stringMethodNames.Contains(method.Name);
+        }
+
+        private static bool IsTryParseMethod(IMethodSymbol method)
+        {
+            return method.Name.StartsWith("TryParse", StringComparison.OrdinalIgnoreCase) &&
+                method.ReturnType.SpecialType == SpecialType.System_Boolean &&
+                method.Parameters.Length >= 2 &&
+                method.Parameters[1].RefKind != RefKind.None;
+        }
```

Comment:
[Aside: I have been occasionally guilty of relying on known TryParse implementation returning default(T) and ignoring the bool in limited scenarios where "or-default" is what I want anyhow, but this is good practice to enforce.]